app/Services/Model/Imp/ScheduleService.php

```diff
@@ -0,0 +1,254 @@
+            $event_ids = $payload['presentation_ids'];
+            foreach ($event_ids as $event_id) {
+                $event = $this->event_repository->getById(intval($event_id));
+                if (!$event instanceof SummitEvent) continue;
+                $schedule_event = $schedule->getScheduledSummitEventByEvent($event);
+                if (is_null($schedule_event)) continue;
+                $filtered_schedule_events[] = $schedule_event;
+            }
+        } else {
+            //filter criteria to promote schedule events by start_date, end_date and location_id
+            $summit = $schedule->getSummit();
+            $start_date = null;
+            $end_date = null;
+            $location = null;
+            if (isset($payload['start_date'])) {
+                $start_date = $summit->parseDateTime(intval($payload['start_date']));
+            }
+            if (isset($payload['end_date'])) {
+                $end_date = $summit->parseDateTime(intval($payload['end_date']));
+            }
+            if (isset($payload['location_id'])) {
+                $location = $summit->getLocation(intval($payload['location_id']));
+            }
+            $filtered_schedule_events =
+                $schedule->getScheduledSummitEventsByLocationAndDateRange($start_date, $end_date, $location);
+        }
+
+        return $filtered_schedule_events;
+    }
+
+    /**
+     *@inheritDoc
+     */
+    public function publishProposedActivityToSource(
+        string $source, int $presentation_id, array $payload):SummitProposedScheduleSummitEvent {
+
+        $schedule = $this->tx_service->transaction(function () use ($source, $presentation_id, $payload) {
+
+            $event = $this->event_repository->getById($presentation_id);
+            if (!$event instanceof SummitEvent)
+                throw new EntityNotFoundException("event id {$presentation_id} does not exists!");
+
+            $schedule = $this->schedule_repository->getBySourceAndSummitId($source, $event->getSummitId());
+
+            if ($event instanceof Presentation) {
+                $selection_plan = $event->getSelectionPlan();
+                if (!$selection_plan instanceof SelectionPlan)
+                    throw new EntityNotFoundException("presentation id {$presentation_id} does not have a selection plan");
+
+                if (!$selection_plan->isAllowProposedSchedules())
+                    throw new ValidationException("selection plan id {$selection_plan->getId()} does not allow proposed schedules");
+            }
+
+            if (is_null($schedule)) {
+                $member = ResourceServerContext::getCurrentUser(false);
+                $schedule = new SummitProposedSchedule();
+                $schedule->setSource($source);
+                $schedule->setSummit($event->getSummit());
+                if (isset($payload['schedule_name'])) {
+                    $schedule->setName($payload['schedule_name']);
```

Comment:
there is no validation rule set at controller level for this field
please add it

Reply:
done